gnmi/gnmi.go

```diff
@@ -0,0 +1,152 @@
+}
+
+func NewGNMIExporter(cfg *Config) (*GNMI, error) {
+	telemSrv, err := lwotgtelem.New(context.Background(), cfg.TargetName)
+	if err != nil {
+		return nil, err
+	}
+
+	gnmiLis, err := net.Listen("tcp", fmt.Sprintf("%s", cfg.Addr))
+	if err != nil {
+		klog.Exitf("cannot listen on %s, err: %v", cfg.Addr, err)
+	}
+
+	var opts []grpc.ServerOption
+
+	if cfg.CertFile != "" {
+		creds, err := credentials.NewServerTLSFromFile(cfg.CertFile, cfg.KeyFile)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create gNMI credentials, %v", err)
+		}
+		opts = append(opts, grpc.Creds(creds))
+	}
+
+	return &GNMI{
+		cfg:      cfg,
+		metricCh: make(chan *pmetric.Metrics, cfg.BufferSize),
+		lis:      gnmiLis,
+		srv:      grpc.NewServer(opts...),
+		telemSrv: telemSrv,
+	}, nil
+}
+
+func (g *GNMI) Start(_ context.Context, _ component.Host) error {
+	fmt.Println("starting")
+	reflection.Register(g.srv)
+	gpb.RegisterGNMIServer(g.srv, g.telemSrv.GNMIServer)
+	go g.srv.Serve(g.lis)
+	g.telemSrv.AddTask(gnmit.Task{
+		Run: g.handleMetrics,
+	})
+	fmt.Println("started")
+	return nil
+}
+
+func (g *GNMI) Stop(_ context.Context) error {
+	close(g.metricCh)
+	g.srv.GracefulStop()
+	return nil
+}
+
+func (g *GNMI) storeMetric(ctx context.Context, md pmetric.Metrics) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		g.metricCh <- &md
+	}
+	return nil
+}
+
+func (g *GNMI) handleMetrics(_ gnmit.Queue, updateFn gnmit.UpdateFn, target string, cleanup func()) error {
+	go func() {
+		var updates []*gpb.Update
+		for ms := range g.metricCh {
+			rms := ms.ResourceMetrics()
+			for i := 0; i < rms.Len(); i++ {
+				rm := rms.At(i)
+				ilms := rm.ScopeMetrics()
+				for j := 0; j < ilms.Len(); j++ {
+					ilm := ilms.At(j)
+					ms := ilm.Metrics()
+					for k := 0; k < ms.Len(); k++ {
+						m := ms.At(k)
+						switch m.Type() {
+						case pmetric.MetricTypeSum:
+							sumMetrics := m.Sum().DataPoints()
+							for l := 0; l < sumMetrics.Len(); l++ {
+								sumMetric := sumMetrics.At(l)
+								updates = append(updates, &gpb.Update{
+									Path: &gpb.Path{
+										Elem:   g.toPathElems(m.Name()),
+										Target: g.cfg.TargetName,
+									},
+									Val: &gpb.TypedValue{
+										Value: &gpb.TypedValue_IntVal{
+											IntVal: sumMetric.IntValue(),
+										},
+									},
+								})
+							}
+						}
+					}
+				}
+			}
+
+			if err := updateFn(&gpb.Notification{
+				Timestamp: time.Now().Unix(),
+				Prefix: &gpb.Path{
+					Target: g.cfg.TargetName,
+					Elem: []*gpb.PathElem{
+						{
+							Name: g.cfg.TargetName,
+						},
+					},
+				},
+				Update: updates,
+			}); err != nil {
+				klog.Errorf("failed to send updates %v", err)
+			}
+		}
+
+	}()
+	return nil
+}
+
+func (g *GNMI) toPathElems(name string) []*gpb.PathElem {
+	var elems []*gpb.PathElem
+	for _, p := range strings.Split(name, g.cfg.Sep) {
+		elems = append(elems, &gpb.PathElem{
+			Name: p,
+			// TODO (alshabib): support keyed paths.
```

Comment:
Does such a keyed path exist? There'd be no need to necessarily have keys in the paths unless we see a need to have them.

Reply:
Yup - agreed. For now, we have no use for this but if we did this is where it could go.